### PR TITLE
KEYCLOAK-11293 Update theme build for keycloak-preview

### PR DESCRIPTION
**Pull request by ASzc**
This changes the themes pom to work with more than one `package.json` file. The
only other one at the moment is for the new account console / account2.

I was not able to use one global node_modules directory, and make it work with
all the requirements, like being able to run the account2 build scripts, and
working with Windows, etc. So, there are two directories, with the possiblity
of multiple copies of a module.

However, I should have saved some megabytes elsewhere, by changing how the
filters run. They used to be done when updating the modules, but now they are
applied at build time, so there's no second copy of the filtered files in git.
This also has the advantage of allowing the filters to affect the modules
globally, and avoids copying them for each package.json.

The documentation has been updated to reflect the new changes.

Since we're building directly out of the source directories, it is possible in
a local dev environment for unintended files (e.g. old compiled .js files),
placed within src/main/resources/, to be included in the themes jar. This
shouldn't be a problem for actual builds though, which use a fresh clone.

Other small changes include refactoring the npm setup stuff to a global
definition, and the introduction of some properties to avoid duplicating path
definitions everywhere.

This commit does not include the churn that would result from running the
update command: `mvn clean -Pnpm-update package` I think it best to keep that
in its own commit.

### CODE REVIEW

**Comment by ssilvert** [comment]
> 3. Produce a tarball of the module and attach it so it is deployed as a maven
>    artifact of org.keycloak:keycloak-themes.
I don't understand this one.  What is the tarball for?  Are we doing this with other themes?

**Comment by ASzc** [comment]
@ssilvert It'd be for any other POMs to consume. It's not necessary if it's used within the themes pom only.

**Comment by ssilvert** [comment]
> @ssilvert It'd be for any other POMs to consume. It's not necessary if it's used within the themes pom only.

I think we should only do it if we are doing it for the other themes.  I actually don't know the answer to that off the top of my head.

**Comment by abstractj** [comment]
@ASzc I created a thread at the mailing list to collect more details to this task. When you get the chance could you please take a look https://lists.jboss.org/pipermail/keycloak-dev/2019-October/012734.html ? Also, I did the updates on the same Jira https://issues.jboss.org/browse/KEYCLOAK-11293

**Comment by ASzc** [comment]
Pushed a reworked version, will update the PR description once I am able to confirm a typescript error it's raising is not due to my changes.

I've given up on trying to make two package.json's share one node_modules directory, it's not possible to do without symlinks, and still be able to run the build script, as far as I can tell.

**Comment by ASzc** [comment]
Updated the commit and description for the reworked changes, please have another look.

The type error in the account2 build still needs to be resolved.

**Comment by abstractj** [comment]
The failures with Typescript should be fixed by https://github.com/keycloak/keycloak/pull/6550. Feel free to just cherry-pick the changes there and incorporate to your PR if you prefer.

**Comment by ASzc** [comment]
Rebased to include Bruno's changes. Also fixed a small error that popped up during the rebase.

I've added a second commit with the result of running `mvn clean -Pnpm-update package`.

**Comment by ssilvert** [comment]
@ASzc Obviously, we first need to make sure the tests are passing before we merge.

Also, it's a little challenging to review this since clicking on the "Files changed" tab above refers to more than 10,000 files.  Is there any way you can help us to see the non-node_modules changes?

**Comment by vmuzikar** [comment]
> Also, it's a little challenging to review this since clicking on the "Files changed" tab above refers to more than 10,000 files. Is there any way you can help us to see the non-node_modules changes?

@ssilvert I usually just check out such PRs to my local machine so I can conveniently go through any relevant changes (filtering out e.g. `node_modules`).

**Comment by abstractj** [comment]
> @ssilvert I usually just check out such PRs to my local machine so I can conveniently go through any relevant changes (filtering out e.g. `node_modules`).

+1 @ssilvert probably this may help:
```
git diff --name-only KEYCLOAK-11293..master  -- . ':!*node_modules*'
```

**Comment by abstractj** [comment]
@ASzc we need to make tests pass on Travis first. It seems that something is missing looking at the logs:
```
[ERROR] Error: Cannot find module '../lib/tsc.js'
[ERROR] Require stack:
[ERROR] - /home/travis/build/keycloak/keycloak/themes/src/main/resources/theme/keycloak-preview/account/resources/node_modules/typescript/bin/tsc
[ERROR]     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:794:15)
[ERROR]     at Function.Module._load (internal/modules/cjs/loader.js:687:27)
[ERROR]     at Module.require (internal/modules/cjs/loader.js:849:19)
[ERROR]     at require (internal/modules/cjs/helpers.js:74:18)
[ERROR]     at Object.<anonymous> (/home/travis/build/keycloak/keycloak/themes/src/main/resources/theme/keycloak-preview/account/resources/node_modules/typescript/bin/tsc:2:1)
[ERROR]     at Module._compile (internal/modules/cjs/loader.js:956:30)
[ERROR]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
[ERROR]     at Module.load (internal/modules/cjs/loader.js:812:32)
[ERROR]     at Function.Module._load (internal/modules/cjs/loader.js:724:14)
[ERROR]     at Function.Module.runMain (internal/modules/cjs/loader.js:1025:10) {
[ERROR]   code: 'MODULE_NOT_FOUND',
[ERROR]   requireStack: [
[ERROR]     '/home/travis/build/keycloak/keycloak/themes/src/main/resources/theme/keycloak-preview/account/resources/node_modules/typescript/bin/tsc'
[ERROR]   ]
[ERROR] }
```

**Comment by ASzc** [comment]
I've put them into two commits, the important changes are in the [first one](https://github.com/keycloak/keycloak/pull/6432/commits/8c5bfcccd569163ec55d904ea586310a6c746a21).

I'll have a look at that error, it seems similar to ones I've seen locally.

**Comment by ASzc** [comment]
Pretty sure the travis error is due to a overly broad .gitignore removing all the js files from the node_modules commit. I've amended the second commit with all the js files. Will see if that fixes it

**Comment by ASzc** [comment]
I'll note that with that gitignore fixed, we now have 30k files changed. Some of those will just be moved files, but many are new because of account2's large dependency tree. Are there any sorts of files in node_modules we don't need for development purposes? I could use the gitignore to filter those out.

**Comment by ASzc** [comment]
Travis is now failing with this error:

`The job exceeded the maximum log length, and has been terminated.`

I think it's because it prints out all the changed files at the beginning?

**Comment by stianst** [comment]
I added some cleanup to the old package.json of files that are needed that
are removed after being pulled from NPM. I think that's a better approach
than .gitignore as it allows you to check things locally before committing.

**Comment by stianst** [comment]
With 30K files we really do need to do something as it is becoming beyond
unmanageable.

I wonder if we really should consider the next step and not commit
node_modules at all? How realistic and how much of a pain would that be?

**Comment by ASzc** [comment]
We'd need to change things in the PNC build config, to build it pretty much the same as we do for the nodejs-connect adapter. This is inherently a wrapper build, so we don't need to wait for native NPM build support in PNC.

Obviously the pom stuff in this PR would need to be redone as well. First build in a fresh clone will be a little bit slower than it is now. I'd try to design it so that the node_modules don't need to be re-downloaded all the time, however it'd also have to prevent them from going out of sync. Maybe this `npm ci` command we talked about before could help there.

**Comment by vmuzikar** [comment]
> I wonder if we really should consider the next step and not commit node_modules at all? How realistic and how much of a pain would that be?

+1000
It'd make a lot of sense to download dependencies during build and not to include them in the repo. However, with the tech preview in mind I'm afraid this is only doable as a follow-up task.

**Comment by abstractj** [comment]
> With 30K files we really do need to do something as it is becoming beyond unmanageable. I wonder if we really should consider the next step and not commit node_modules at all? How realistic and how much of a pain would that be?

@stianst +1

@ASzc do you have any ballpark estimate? Only to have an idea how hard it would be.

**Comment by ASzc** [comment]
@abstractj I don't think it'd be too hard. One way of doing it would be to:

1. Change the pom to remove the `npm-update` profile and make those actions (`npm install`) the default.
2. Use `.gitignore` to prevent any `node_modules` directory from being committed.
3. To avoid stale modules, either re-enable the `package-lock.json`s or use exact versions in the `package.json`s.

Uncertainty would be all in step 3, but I think I can have the change ready to review in a day or two. How long review takes is another matter.

**Comment by ssilvert** [comment]
> With 30K files we really do need to do something as it is becoming beyond unmanageable. I wonder if we really should consider the next step and not commit node_modules at all? How realistic and how much of a pain would that be?

The thing I'm spiking right now takes each runtime dependency and rolls it up into a single file.  It will end up being something like 10 to 12 files.  So if we can say that these "rolled up" files qualify as our source then maybe we could just check those into GitHub?

**Comment by vmuzikar** [comment]
@ssilvert Sure, 12 files sounds better than 30k but still... committing no dependencies at all would be the cleanest and most future proof solution? I mean, normally we never ever commit deps (maven, npm_modules in node.js adapter, ...), right? It'd be nice if we could do it here as well. Not saying it needs to be done for the tech preview, though.

**Comment by ssilvert** [comment]
> @ssilvert Sure, 12 files sounds better than 30k but still... committing no dependencies at all would be the cleanest and most future proof solution? I mean, normally we never ever commit deps (maven, npm_modules in node.js adapter, ...), right? It'd be nice if we could do it here as well. Not saying it needs to be done for the tech preview, though.

Yes, committing no dependencies is our ultimate goal.  But we can't do that right now because of our reproduceability requirements.  We can't pull from a public npm repo to do a production build.  The PNC project is supposed to solve this but as far as I know we are still waiting.

**Comment by ASzc** [comment]
@ssilvert I think we can do this now. Our process for nodejs-connect has been semi-formalized in recent months, using the NPM proxy which PNC deployed. This proxy is what native NPM build support will use when that's rolled out, so it's official and supported. Internal docs [here](https://url.corp.redhat.com/b29ed5f).

 I've completed an [alternative branch here](https://github.com/ASzc/keycloak/commits/KEYCLOAK-11293_dynamic), and I can swap it in for this PR if people want it.

**Comment by ssilvert** [comment]
> @ssilvert I think we can do this now. Our process for nodejs-connect has been semi-formalized in recent months, using the NPM proxy which PNC deployed. This proxy is what native NPM build support will use when that's rolled out, so it's official and supported. Internal docs [here](https://url.corp.redhat.com/b29ed5f).
> 
> I've completed an [alternative branch here](https://github.com/ASzc/keycloak/commits/KEYCLOAK-11293_dynamic), and I can swap it in for this PR if people want it.

I think that's a good idea.  +1 for the swap.

But let's make sure that this is truly official and supported.  @stianst wdyt?

**Comment by vmuzikar** [comment]
@ASzc +1 for swapping this for the alternative branch. Or perhaps create another PR from the alternative branch just in case...

One thing I'm a bit concerned about it that we would be touching not only deps for the new Account Console but all npm deps we use in general. E.g. currently it's broken as it replaces all the deps in `common/resources/node_modules` with empty `undefined.js` files. Maybe should this PR focus only on the new Account Console and leave other deps as they are just for now?

**Comment by abstractj** [comment]
@ASzc I got the chance to talk with @stianst today and he's aboard with the idea of adding those commits, as long as we don't have any blockers with PNC and it's doable for 7.4.
Based on all the comments above, I'd say go for it.

**Comment by ASzc** [comment]
Hope to make the suggested change to keep the old modules the same. Will have time after my presentation tomorrow.

**Comment by ASzc** [comment]
Have made the change and swapped the branches. Had to do a bit of a hybrid, but the old modules will still only be updated if you run `-Pnpm-update`.

Branch now has an extra commit with the churn too. Should be good to merge, pending final review.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Thanks @ASzc for the update!

Unfortunately, I've found two issues.

1. When the project is build from scratch (freshly checked out repo), the JS files transpiled from TypeScript are missing from the themes JAR (and from distribution as well of course). It seems like the JAR is created before transpiling TypeScript.
2. `package-lock.json` was deleted. I think we should leave it there.

**Comment by ASzc** [comment]
Will have a look

**Comment by ASzc** [comment]
Should be good now.

I put package-lock generation back in:

```
aszczucz@tyrfing ~/s/k/keycloak2 (KEYCLOAK-11293_dynamic)> 
find themes/ -name '*package-lock*' | grep -v 'node_modules'
themes/target/classes/theme/keycloak-preview/account/resources/package-lock.json
themes/src/main/resources/theme/keycloak-preview/account/resources/package-lock.json
```

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@ASzc Thanks for the update! Maybe I'm doing something wrong but I think I've found another bug. I don't know why but when I build distribution from a cleanly checked out repo, new account console doesn't work because it's missing `account/keycloak-preview/node_modules/@patternfly/react-core/dist/umd/index.js`.

It's most probably cause by this:
https://github.com/keycloak/keycloak/blob/e059d8b11f558523727e8554aa2b4558434e2c48/themes/pom.xml#L111

**Comment by ASzc** [comment]
To be safe, I included all `index.js` files under `@patternfly`

File: themes/pom.xml
Comment (vmuzikar) [inline themes/pom.xml]: The `--no-shrinkwrap` option causes that our package lock is ignored and regular package.json is used instead. That means we e.g. install newer dependencies than we expect (because we use `^` when specifying versions in package.json). And because of the newer dependencies the Account Console is broken. I know this is partially an issue with our package.json (we probably shouldn't use `^`) but at the same time I believe it's not a good practice to effectively ignore package lock.

Or did I miss something, do we need `--no-shrinkwrap` for some reason?

File: themes/pom.xml
Comment (vmuzikar) [inline themes/pom.xml]: I'd remove this if possible. It seems more dependencies (not just patternfly/react) rely on an `index.js` (e.g. exenv) and without it the Account Console doesn't work.

File: themes/pom.xml
Comment (vmuzikar) [inline themes/pom.xml]: This can be removed too if we remove `<exclude>**/index.js</exclude>`.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
I did some more testing. Unfortunately, I've found two more issues (both break the new Account Console). This really should be last issues – seems like if we resolve them, we should be good to go.

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: It was part of the options that Stian added. I thought we were going to specify exact versions in package.json? If `npm install` is called every time maven runs, we don't want npm introducing noise into the commit with it randomly changing package-lock.json all the time.

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: Including all the index.js files will introduce about 13M into the distribution, these aren't super light files for some reason.

File: themes/pom.xml
Comment (vmuzikar) [reply themes/pom.xml]: @ASzc I see. But shouldn't package lock also help us with transitive dependencies versions?

What about using [`npm ci`](https://blog.npmjs.org/post/171556855892/introducing-npm-ci-for-faster-more-reliable) instead of `npm install`? It seems it's exactly what we need. It uses package lock to install dependencies, doesn't update it and produce "predictable" builds.

File: themes/pom.xml
Comment (vmuzikar) [reply themes/pom.xml]: Yes, but this way we'll need to manually pick those index.js files that are really required by some of our (transitive) dependencies. And that is IMHO quite a big hassle, mainly when e.g. updating dependencies or introducing new ones. Every time we'll need to manually investigate whether this or that index.js is really required.

Comment (vmuzikar) [reply themes/pom.xml]: Thinking about this more. Maybe this is more of a generic question what to really include in the distribution. Couldn't this be somehow derived from [systemjs.config.js](https://github.com/keycloak/keycloak/blob/master/themes/src/main/resources/theme/keycloak-preview/account/resources/systemjs.config.js)? @ssilvert?

npm surely downloads lots of "garbage" and we need only a few files.

On the other hand this is basically an optimization task. Couldn't this be perhaps done after the tech preview?

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: The main disadvantage of `npm ci` is that it deletes the node_modules directory each time it's called, so some external caching would be needed to avoid a complete slowdown. I think npm does do this, in `~/.npm`, so I can swap the command over to `ci`

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: We can do the optimization later, yeah.

**Comment by ASzc** [comment]
- Removed the exclusion for `index.js` entirely
- Swapped `npm install` to `npm ci`
- Removed `--no-shrinkwrap` from the install arguments

I'm now seeing ~3000 files being included in the `Run mvn clean package -Pnpm-update` commit, rather than ~1000, otherwise seems ok.

File: themes/pom.xml
Comment (vmuzikar) [inline themes/pom.xml]: We should add `package-lock.json` as well. It's currently included in the distribution.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@ASzc Thanks for the update.

I did some more testing and investigating and I can confirm the Account Console build finally works! :)
I've found some more issue but all of them are just polishing.

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: I'll put it back, but do we want `--no-shrinkwrap --no-package-lock` back in as well? Those will disable package-lock.

Comment (vmuzikar) [reply themes/pom.xml]: @ASzc The common deps have no package lock (which is not ideal as it e.g. prevent us from easily scanning for vulnerabilities) but it was like it before and this PR isn't changing that. So yes please, leave the argument here exactly as they were.

**Comment by ASzc** [comment]
- Updated docs
- Filtered package-lock from distribution
- `-Pnpm-update` now runs Stian's arguments verbatim for the old modules
- New modules remain on `npm ci` with a package-lock used/generated.

Back down to ~1000 files

**Comment by ASzc** [comment]
Travis error seems to be a random test failure.

**Review by vmuzikar — DISMISSED** [review]
@ASzc From my PoV looks good! Thank you for your patience with my neverending requests. :D

**Comment by martin-kanis** [comment]
@ASzc I checked the PNC build and its output and it looks good to me. However I'm wondering if you could run it from the pipeline? I would like to verify whether the pipeline can transform prod-arguments.json to a correct build configuration.

**Comment by ASzc** [comment]
@martin-kanis Thanks for looking. The stuff added to prod-arguments in this PR is based on the prod-arguments of [the nodejs adapter](https://github.com/keycloak/keycloak-nodejs-connect/blob/master/product/prod-arguments.json), so it should work. The current build config script always selects an NPM-containing environment now, so that should be ok too.

Is merge monster up to verifying my assumptions?

File: themes/pom.xml
Comment (douglaspalmer) [inline themes/pom.xml]: Do we need to override the version from the parent pom for this plugin? Can we not just change the version in the parent? Also, is the a feature or bug fix in 1.8.0 that we need? The newest version in brew is currently 1.7.6.

File: themes/pom.xml
Comment (douglaspalmer) [inline themes/pom.xml]: If you use ${basedir} instead of node here you will only generate one node directory instead of node/node/

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: Parent POM specifies 1.5, hadn't seen that. I don't know of a specific reason to use later than that version, but since we've been using 1.8.0 [since before my PR](https://github.com/keycloak/keycloak/blob/89fc0e1fcc3df43e4a8e89c817b897f8aa9a5d60/themes/pom.xml#L141), I don't want to change that now. I'll update the parent POM's version, since the plugin isn't used anywhere but themes.

The nice thing about building in PNC is it will import the third party artifacts for us, unlike Brew.

File: themes/pom.xml
Comment (ASzc) [reply themes/pom.xml]: I'll try to implement this when I do the version definition move.

Comment (ASzc) [reply themes/pom.xml]: Seems to work. Anyone with an existing node/node will need to manually delete that directory, but works fine from clean.

**Comment by ASzc** [comment]
- Rebased
- Moved frontend plugin version definition to keycloak parent pom, remains 1.8.0
- Changed node install directory to be slightly cleaner

**Review by douglaspalmer — DISMISSED** [review]
LGTM

**Comment by stianst** [comment]
As we now don't include node_modules for account2 in the repository and download that at build-time I'd say it would be better to do the same for the old stuff as well. I appreciate that there is a slight risk of regression there, but should be fairly easy to compare a build before and after the changes to check that the included files match.

**Comment by stianst** [comment]
In a distribution in "themes/keycloak-preview/account/resources" I now see node (37M) and node_modules (221M). That is not acceptable I'm afraid, as just including account2 now increases the distribution size with 259M. So there's a lot of files to filter out here.

**Comment by stianst** [comment]
Unzip of master is 554M (310M is themes directory), while 8.0.1 is 309M.

**Comment by stianst** [comment]
We can probably also filter quite a bit in the old node_modules as well. Just patternfly and rcue are ~40M.

**Comment by stianst** [comment]
It's quite crazy the amount of dependencies that goes into themes/keycloak-preview/account/resources/node_modules. 412 by counting directories. That is not going to be sustainable from disc size or from CVE perspective. We simply can't maintain that amount of dependencies. Are all these really required by the new account console?!? Or is there a large amount of optional/transitive/dev dependencies being pulled-in?

**Review by stianst — CHANGES_REQUESTED** [review]
This PR can't be merged in it's current state due to the increased size of the distribution, the enormous amount of new dependencies, and it is far from ideal that the new and the old is being handled differently.

**Comment by ssilvert** [comment]
> It's quite crazy the amount of dependencies that goes into themes/keycloak-preview/account/resources/node_modules. 412 by counting directories. That is not going to be sustainable from disc size or from CVE perspective. We simply can't maintain that amount of dependencies. Are all these really required by the new account console?!? Or is there a large amount of optional/transitive/dev dependencies being pulled-in?

It's definitely the latter.

**Comment by vmuzikar** [comment]
> As we now don't include node_modules for account2 in the repository and download that at build-time I'd say it would be better to do the same for the old stuff as well. I appreciate that there is a slight risk of regression there, but should be fairly easy to compare a build before and after the changes to check that the included files match.

@stianst This is [my "fault"](https://github.com/keycloak/keycloak/pull/6432#issuecomment-563197841). ;) At some point we had the npm stuff unified for both consoles. However, touching Admin Console deps is always quite error prone (which proved to be true here too). Also it was a bit out of scope of this PR (building the new Account Console), so we agreed on not touching the old dependencies for Admin Console just to be on the safe site for now.



> It's quite crazy the amount of dependencies that goes into themes/keycloak-preview/account/resources/node_modules. 412 by counting directories. That is not going to be sustainable from disc size or from CVE perspective. We simply can't maintain that amount of dependencies. Are all these really required by the new account console?!? Or is there a large amount of optional/transitive/dev dependencies being pulled-in?

We originally wanted to do the optimization stuff later as it's basically a separate task – see https://github.com/keycloak/keycloak/pull/6432#discussion_r364797592. IMHO people need to expect that a tech preview is not 100% optimized (which in our case is not only the npm stuff but also e.g. network overhead). Maybe we could implement some temporary solution just for the tech preview?

**Comment by ASzc** [comment]
@stianst You'll want to run `git clean -idx` or otherwise use a clean clone to get an accurate picture of size. Since we're building out the source directory, extra stuff can sneak in (like that node directory) that isn't part of the jar in a real build.

I see a growth in the unpacked themes jar from 51 MB to 265 MB. Still a pretty big increase.

I think `npm ci` may force the download of optional dependencies. I'll see if there's a way to filter those out from the jar. If people can suggest some other filters to add, I can put them in at the same time.

**Comment by ASzc** [comment]
If I insert a `npm prune --production` into the build after the typescript compilation, the unpacked themes size is reduced to 194 MB.

It may be undesirable for this prune to run all the time though. If so, I can maybe move it into a profile that isn't normally active?

**Comment by vmuzikar** [comment]
@ASzc I'm afraid we'll have to go through [systemjs.config.js](https://github.com/keycloak/keycloak/blob/master/themes/src/main/resources/theme/keycloak-preview/account/resources/systemjs.config.js) and add filters based on that. But we need some way to automate this after the tech preview.

**Comment by stianst** [comment]
How many dependencies are we including? The number of dependencies will be important from a CVE perspective.

194 MB is that in addition, or total? I'm still far from happy with that as it increases the download and installation size significantly.

I'm okay with leaving the old modules as is for 7.4. This would have to be addressed after 7.4 though.

**Comment by vmuzikar** [comment]
> How many dependencies are we including? The number of dependencies will be important from a CVE perspective.


BTW, regarding CVEs. We should handle the deps reduction on npm level as much as possible. IIRC, basically any vulnerability scanner relies on package-lock.json. If we use e.g. Maven for "selecting" the deps, those scanners would return some "false positives" for libraries we don't even use which then results in deeper manual investigation than required otherwise. But IMHO this could be done after 7.4. For 7.4 we could use some more

> 194 MB is that in addition, or total? I'm still far from happy with that as it increases the download and installation size significantly.

Assuming from https://github.com/keycloak/keycloak/pull/6432#issuecomment-576342245, I believe it's final size of themes directory in the dist.

> I'm okay with leaving the old modules as is for 7.4. This would have to be addressed after 7.4 though.

Created https://issues.redhat.com/browse/KEYCLOAK-12740

**Comment by ASzc** [comment]
From meeting: For now, we'll narrow the files down manually with a filter list, to just the ones actually needed by systemjs.config.js

**Comment by ASzc** [comment]
With @vmuzikar's help I've generated a minimal list of includes that should serve account2 without any cruft. The new unpacked size of the themes jar is only 55 MB (!) compared to 51 MB without account2, so an increase of ~4 MB.

I think this should satisfy you @stianst :)

The process of creating the list involves running account2 in your browser, so it's not ideal for the long term. I've documented it in a comment in the POM.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
The following files are missing in the build:
```
@patternfly/patternfly/assets/fonts/webfonts/fa-solid-900.ttf
@patternfly/patternfly/assets/fonts/webfonts/fa-solid-900.woff
@patternfly/patternfly/assets/fonts/webfonts/fa-solid-900.ttf
```
@ASzc Could you please add them to the filters? It was not apparent at first sight as they're only required by Welcome Screen and only when i18n is enabled in the realm.

**Comment by ASzc** [comment]
Added those three

**Review by vmuzikar — APPROVED** [review]
Everything LGTM, thank you @ASzc!

**Review by douglaspalmer — APPROVED** [review]
LGTM

**Review by stianst — APPROVED** [review]

**Pull request merged**
